### PR TITLE
Fix runc crushes when parsing invalid JSON specification file

### PR DESCRIPTION
**Pull request by AITsygunka**
Fixes [#3727](https://github.com/opencontainers/runc/issues/3727#issue-1573797885)

### CODE REVIEW

File: spec.go [hunk 1]
```diff
@@ -123,8 +123,12 @@ func loadSpec(cPath string) (spec *specs.Spec, err error) {
...
 
-	if err = json.NewDecoder(cf).Decode(&spec); err != nil {
-		return nil, err
+	if err = json.NewDecoder(cf).Decode(&spec); err != nil || spec == nil {
+		if err == nil {
+			return nil, fmt.Errorf("Invalid JSON specification file %s format", cPath)
+		} else {
+			return nil, err
+		}
 	}
```
Comment (cyphar) [inline spec.go:132]:
```suggestion
	if err = json.NewDecoder(cf).Decode(&spec); err != nil || spec == nil {
		return nil, err
	} else if spec == nil {
		return nil, fmt.Errorf("config cannot be null")
	}
```

It should be noted the issue here is not with invalid JSON (that would return an error from `json.Decode`), but instead with a `config.json` that is set to `null` (which is valid JSON). I feel like another option would be to have `spec` be `specs.Spec` which would stop it from being able to be `nil`.

**Comment by cyphar** [comment]
You need to add a `Signed-off-by:` line to your commit(s) which indicates that you attest the [Developer Certificate of Origin](https://developercertificate.org/) a statement about your contributions that you *must read* before signing (don't worry, it's quite short and easy-to-read). You can add it to your commits with `git commit --amend -s`, and then doing a `git push --force`.

###### **NOTE**: This is a saved reply. Sorry if it reads as a cookie-cutter response, it was written so that newcomers understand what a "DCO" is and make the process for contributing a little less scary. ######

File: spec.go [hunk 2]
```diff
@@ -125,6 +125,8 @@ func loadSpec(cPath string) (spec *specs.Spec, err error) {
 
 	if err = json.NewDecoder(cf).Decode(&spec); err != nil {
 		return nil, err
+	} else if spec == nil {
```
Comment (kolyshkin) [inline spec.go:128]: `else` is redundant here because of the return statement above it.

File: spec.go [hunk 3]
```diff
@@ -125,6 +125,8 @@ func loadSpec(cPath string) (spec *specs.Spec, err error) {
...
+		return nil, fmt.Errorf("config cannot be null")
```
Comment (kolyshkin) [inline spec.go:129]: Nit: hen there are no %-style arguments, `errors.New` is better.

**Review by kolyshkin — CHANGES_REQUESTED** [review]

**Review by kolyshkin — DISMISSED** [review]
LGTM

**Comment by kolyshkin** [comment]
CI failure is unrelated; it was recently reported as #3742

**Comment by kolyshkin** [comment]
@AITsygunka please rebase to fix CI

**Review by AkihiroSuda — APPROVED** [review]

**Review by kolyshkin — APPROVED** [review]
LGTM

**Pull request merged**